File: apps/cargo-tracker/src/Cargotracker/Estimation/Domain/Model/Value/EstimateStatus.hs
{- | 見積状態 値オブジェクト (US01, IT2)

- Created: 見積発行直後 (有効)
- Expired: 期限切れ (再見積が必要)

IT2 では作成時 Created のみ。Expired への遷移ロジックは
IT3 で発行時刻 + 有効期間ポリシー導入時に追加する。
-}
module Cargotracker.Estimation.Domain.Model.Value.EstimateStatus
  ( EstimateStatus (..),
    estimateStatusToText,
    parseEstimateStatus,
  ) where

import Data.Text (Text)

import Cargotracker.Shared.Domain.DomainError (DomainError (..))

data EstimateStatus
  = Created
  | Expired
  deriving stock (Eq, Show, Enum, Bounded)

estimateStatusToText :: EstimateStatus -> Text
estimateStatusToText Created = "Created"
estimateStatusToText Expired = "Expired"

parseEstimateStatus :: Text -> Either DomainError EstimateStatus
parseEstimateStatus "Created" = Right Created
parseEstimateStatus "Expired" = Right Expired
parseEstimateStatus other =
  Left (InvalidBookingId ("unknown EstimateStatus: " <> other))

File: docs/article/source/java-3/apps/backend/bookingms/src/main/java/com/example/bookingms/domain/model/valueobjects/EstimateStatus.java
package com.example.bookingms.domain.model.valueobjects;

/**
 * 見積の状態（US01）。
 *
 * <p><strong>本 IT で起こす遷移は無い。</strong>作成した見積はそのまま残る。
 * それでも 2 値を宣言するのは正典に合わせるためであり、期限切れの扱い
 * （いつ・誰が {@code EXPIRED} にするか）を決めるのは別のストーリーである
 * ——<strong>書き込む相手がいない状態を先に作らない</strong>。
 */
public enum EstimateStatus {

    /** 作成済み。 */
    CREATED("作成済"),

    /** 期限切れ。**本 IT では誰も遷移させない。** */
    EXPIRED("期限切れ");

    private final String label;

    EstimateStatus(String label) {
        this.label = label;
    }

    /** 画面に出す名前。 */
    public String label() {
        return label;
    }
}
